### ISSUE TITLE
Specifying "custom user counter flags" for the usual timers

### ISSUE THREAD

**Issue by HDembinski** [google/benchmark#830]
I am successfully using Google Benchmark in [Boost.Histogram](https://github.com/boostorg/histogram), but I seem to have run a situation that I cannot solve with the current interface. I would like to have a way to set the number of iterations added per loop in the benchmark.

Here is the situation. I have these two types of benchmarks:
```
static void BenchmarkA(benchmark::State& state) {
  value_generator gen;
  for (auto _ : state)
    function_which_accepts_one_element(gen());
}

static void BenchmarkB(benchmark::State& state) {
  value_generator gen;
  for (auto _ : state)
    function_which_accepts_range_of_elements(gen.begin(), gen.end());
}
```

The second benchmark tests a function that accepts a range of inputs instead of one input at a time, otherwise they are very similar, so I would like to compare the time needed to process one input from the generator in both cases. I propose a new method on `State` which allows me to manually increment the iteration counter by more than one unit per iteration, like so:

```
// requested feature
static void BenchmarkB(benchmark::State& state) {
  value_generator gen;
  for (auto _ : state) {
    function_which_accepts_range_of_elements(gen.begin(), gen.end());
    state.AddIterations(gen.size());
  }
}
```

**Comment by dominichamon** [comment]
On Tue, Jun 11, 2019 at 7:10 AM Hans Dembinski <notifications@github.com>
wrote:

> I am successfully using Google Benchmark in Boost.Histogram
> <https://github.com/boostorg/histogram>, but I seem to have run a
> situation that I cannot solve with the current interface. I would like to
> have a way to set the number of iterations added per loop in the benchmark.
>
> Here is the situation. I have these two types of benchmarks:
>
> static void BenchmarkA(benchmark::State& state) {
>   value_generator gen;
>   for (auto _ : state)
>     function_which_accepts_one_element(gen());
> }
>
> static void BenchmarkB(benchmark::State& state) {
>   value_generator gen;
>   for (auto _ : state)
>     function_which_accepts_range_of_elements(gen.begin(), gen.end());
> }
>
> The second benchmark tests a function that accepts a range of inputs
> instead of one input at a time, otherwise they are very similar, so I would
> like to compare the time needed to process one input from the generator in
> both cases. I propose a new method on State which allows me to manually
> increment the iteration counter by more than one unit per iteration, like
> so:
>
> // requested feature
> static void BenchmarkB(benchmark::State& state) {
>   value_generator gen;
>   for (auto _ : state) {
>     function_which_accepts_range_of_elements(gen.begin(), gen.end());
>     state.AddIterations(gen.size());
>   }
> }
>
> I'm not sure this will do what you want. You don't want to increase the
apparent iterations, you want to report on the throughput of your function.
You can do that today with `state.SetItemsProcessed(gen.size() *
state.iterations())` which should be called after the loop. You can also,
if you care about the size of `value_generator`, use `SetBytesProcessed`
directly (
https://github.com/google/benchmark/blob/master/test/benchmark_test.cc#L93).

**Comment by HDembinski** [comment]
Thank you for the super-fast response! It is true that `SetItemsProcessed` kinda works, but I still would like to see the time per item processed, while `SetItemsProcessed` gives me the inverse, a "X MB/s" rate. This is also interesting, but I am really more interested in the average ns per item.

**Comment by HDembinski** [comment]
Maybe a new flag could be added to toggle this in the report?
`BENCHMARK(my_benchmark)->ShowTimePerItemProcessed()`
This would be backward-compatible.

**Comment by LebedevRI** [comment]
> the time per item processed
Closest you can do is:
```
state.counters["ItemsPerSec"] = Counter((gen.size(), benchmark::Counter::kIsIterationInvariantRate);
```
For the inverted value see https://github.com/google/benchmark/issues/781

**Comment by HDembinski** [comment]
Yes, #781 is related, but it would modify the way how the user counter column is reported. I would like to modify the values displayed in the cpu-time and real-time columns.

**Comment by LebedevRI** [comment]
Custom user counters exist for this very reason, i'm not sure it's worthwhile changing the entire time/cpu columns.

**Comment by HDembinski** [comment]
I propose this, because I think that the time/cpu columns currently do not provide useful information when you process several items per iteration. I think the only time that makes sense to report is the time used per item. This should have been the behavior when the user sets `SetItemsProcessed`, but it isn't. The behavior cannot be changed now, but it is possible to add a new flag `->ShowTimePerItemProcessed()` to get the new behavior.

**Comment by LebedevRI** [comment]
Please can you explain for what particular reason you don't like the custom user counter-based solution?

**Comment by HDembinski** [comment]
Ok, to repeat: I think the time/cpu columns currently do not provide meaningful information when you use `SetItemProcessed`. With solution #781, I can either see the MB/s or the time per item in the user-counter column. Wouldn't it be even better to see both at the same time? time per item in the cpu-time and real-time columns and MB/s in the user-counter column? That's what you would get with the solution I am proposing.

**Comment by HDembinski** [comment]
Also, strictly speaking #781 should give s/MB, while I am interested in ns per item processed.

**Comment by LebedevRI** [comment]
> With solution #781, I can either see the MB/s or the time per item in the user-counter column. 

Why do you think that?
You can have any number of those user-counters.

> Also, strictly speaking #781 should give s/MB, while I am interested in ns per item processed.

Uh, not really?
You already can have `items/s` and `mb/s`, that will only add a bit to flip the entire field,
so you will be able to have both the `s/item` and `s/mb`.

**Comment by HDembinski** [comment]
Ok, you know more than me about user counters, but I am not saying that #781 is bad, but I simply care more about having meaningful values reported for cpu-time and real-time for the reasons that I have already given and which are not invalided by anything you said.

**Comment by dominichamon** [comment]
The reason the library exists is to provide aggregate benchmarks over large
numbers of iterations of some functionality. Ie, cputime and realtime are
meaningful but for a different meaning than you're looking for.

If you're interested in understanding how your function scales with gen
size, then perhaps initialize gen from a range variable. Each of these
sizes will then be reported as a different line item in the output,
allowing you to understand the scaling properties.

If you're interested in the throughput, as above, then
SetItemsProcessed/SetBytesProcessed is the thing you want.

If you're interested in the inverse of that, then we get into user counters.

The core functionality of the library doesn't need to change as we already
provide different options for getting what users might want in different
circumstances, and what you're currently looking for is not the core
use-case.

**Comment by LebedevRI** [comment]
> If you're interested in understanding how your function scales with gen size, then perhaps initialize gen from a range variable. Each of these sizes will then be reported as a different line item in the output, allowing you to understand the scaling properties.

Oh yeah, i've missed the elephant in the room.

**Comment by HDembinski** [comment]
> If you're interested in understanding how your function scales with gen size, then perhaps initialize gen from a range variable. Each of these sizes will then be reported as a different line item in the output, allowing you to understand the scaling properties.

I am not interested in the scaling properties with number of items processed, I am merely interested in the time per item processed. So I think that the solution you are proposing doesn't fit my use case.

If you look back into my motivating example: I want to compare two different forms of interface. In the first case, the default case, the loop over the input is external. In the second case, the loop over the input is internal. In the second case, I can do certain optimizations, certain setup costs are paid only once per many items processed, while they are paid for every item in the first case. Still, in both cases I care about the average time per item processed.

**Comment by dominichamon** [comment]
Is that not just the inverse of items per second? I know we don't report
that directly but you can get at it really quickly, output as csv, and
calculate the reciprocal (which is going to be quicker than getting changes
plumbed through the library).

**Comment by HDembinski** [comment]
> Is that not just the inverse of items per second?

Yes, I can change everything on my end. I can even modify the json-reporter that I am using in my fork and be done with it, but I thought this was an opportunity to add a useful feature upstream.

**Comment by HDembinski** [comment]
I would be willing to contribute code. I am a Boost author, so can contribute something of quality, but if there is no interest then I will find another way to solve my problem.

**Comment by LebedevRI** [comment]
(i personally can only again point at user counters + #781)

**Comment by LebedevRI** [comment]
Fixed by #850. (github didn't parse multiple issues at once..)
